src/handlers/rotate.rs

```diff
@@ -0,0 +1,151 @@
+};
+use axum::extract::{Query, State};
+use axum::headers::authorization::Bearer;
+use axum::headers::Authorization;
+use axum::http::StatusCode;
+use axum::TypedHeader;
+use libvips::{ops, VipsImage};
+use serde::Deserialize;
+use std::fs::rename;
+use uuid::Uuid;
+
+#[derive(Deserialize)]
+pub struct RotateQuery {
+    id: Uuid,
+    angle: f64,
+}
+
+pub async fn rotate_handler(
+    State(server_state): State<ServerState>,
+    TypedHeader(authorization): TypedHeader<Authorization<Bearer>>,
+    query: Query<RotateQuery>,
+) -> Result<String, (StatusCode, String)> {
+    check_api_key(authorization, &server_state.api_key_hash)?;
+
+    if query.angle < 0.0 || query.angle > 360.0 {
+        return Err((
+            StatusCode::BAD_REQUEST,
+            "Angle must be 0.0 < angle < 360.0!".to_owned(),
+        ));
+    }
+
+    let image_directory = determine_img_dir(query.id, ImageSearchBehaviour::Valid);
+    if image_directory.is_err() {
+        return Err((StatusCode::NOT_FOUND, "Image not found!".to_owned()));
+    }
+
+    let image_directory = image_directory.unwrap();
+    let image_directory = image_directory.to_str().unwrap();
+
+    let image_path = determine_img_path(image_directory, query.id);
+    if image_path.is_err() {
+        log::warn!(
+            "Image not found where the path was previously determined. Id: {:?}, Directory: {:?}, Error: {:?}",
+            query.id,
+            image_directory,
+            image_path.err()
+        );
+        return Err((StatusCode::NOT_FOUND, "Image not found!".to_owned()));
+    }
+
+    let image_path = image_path.unwrap();
+    let image_path = image_path.to_str().unwrap();
+
+    let raw_image = match std::fs::read(image_path) {
+        Ok(raw_image) => raw_image,
+        Err(err) => {
+            log::error!(
+                "Error while reading image. Id: {:?}, Error: {:?}, Path: {:?}",
+                query.id,
+                err,
+                image_path
+            );
+            return Err((
+                StatusCode::INTERNAL_SERVER_ERROR,
+                "Error while reading image!".to_owned(),
+            ));
+        }
+    };
+
+    // We need to fully load the image into memory here, as saving a VipsImage to a file it was
+    // memory mapped from is a bad idea. The best solution would be to use `vips_image_new_from_file`
+    // and set memory argument to true (to force loading into memory), but that's not possible
+    // with the rust bindings.
+    let image = match VipsImage::new_from_buffer(raw_image.as_slice(), "") {
+        Ok(image) => image,
+        Err(err) => {
+            log::error!(
+                "Error while opening image. Id: {:?}, Error: {:?}, Path: {:?}",
+                query.id,
+                err,
+                image_path
+            );
+            return Err((
+                StatusCode::INTERNAL_SERVER_ERROR,
+                "Error while opening image!".to_owned(),
+            ));
+        }
+    };
```

Comment:
Can we not just read the image from file now that we do not replace it in place?

Reply:
Fixed